src/lib/util/pair_tests.c

```diff
@@ -0,0 +1,1205 @@
+	{ .attr = FR_TEST_TLV_ROOT, .parent = NULL, .da = &attr_test_tlv_root, .name = "Test-TLV-Root", .type = FR_TYPE_TLV, },
+	{ .attr = FR_TEST_TLV_STRING, .parent = &attr_test_tlv_root, .da = &attr_test_tlv_string, .name = "Test-TLV-String", .type = FR_TYPE_STRING, },
+	{ -1, NULL, NULL, NULL, -1, },
+};
+
+/*
+ *	It will be called before of unit tests.
+ */
+static int load_attr_pairs(fr_pair_list_t *out)
+{
+	fr_dict_adhoc_attr_t *p;
+
+	fr_pair_list_init(out);
+
+	for (p = test_dict_attrs;
+	     p->attr != -1;
+	     p++) if (fr_pair_add_by_da(autofree, NULL, out, *p->da) < 0) return -1;
+
+	return 0;
+}
+
+static int init_adhoc_attrs(fr_dict_adhoc_attr_t *dict_adhoc)
+{
+	fr_dict_adhoc_attr_t *ctx;
+	fr_dict_attr_flags_t dict_flags = {
+		.is_root = true
+	};
+
+	for (ctx = dict_adhoc;
+	     ctx->attr != -1;
+	     ctx++) {
+		fr_dict_attr_t const *parent = ctx->parent ? *ctx->parent : fr_dict_root(dict_test);
+		fr_dict_attr_t const *attr;
+
+		if (fr_dict_attr_add(dict_test, parent, ctx->name, ctx->attr, ctx->type, &dict_flags) < 0) return -1;
+
+		attr = fr_dict_attr_by_name(NULL, parent, ctx->name);
+		if (!attr) return -1;
+
+		/* Set the VALUES */
+		if (ctx->values) {
+			fr_dict_adhoc_attr_value_t *v;
+
+			for (v = ctx->values;
+			     v->key != NULL;
+			     v++) fr_dict_attr_enum_add_name(fr_dict_attr_unconst(attr), v->key, &v->val, false, false);
+		}
+
+		*ctx->da = attr;
+	}
+
+	return 0;
+}
+
+static void pair_tests_init(void)
+{
+	autofree = talloc_autofree_context();
+	if (!autofree) {
+	error:
+		fr_perror("pair_tests");
+		fr_exit_now(EXIT_FAILURE);
+	}
+
+	/*
+	 *	Mismatch between the binary and the libraries it depends on
+	 */
+	if (fr_check_lib_magic(RADIUSD_MAGIC_NUMBER) < 0) goto error;
+
+	if (!fr_dict_global_ctx_init(autofree, dict_dir)) goto error;
+
+	/*
+	 *	Set the root name of the dictionary
+	 */
+	dict_test = fr_dict_alloc("test", 666);
+	if (!dict_test) goto error;
+
+	if (init_adhoc_attrs(test_dict_attrs) < 0) goto error;
+
+	/* Initialize the "sample_pairs" list */
+	if (load_attr_pairs(&sample_pairs) < 0) goto error;
+}
+
+/*
+ *	Tests functions
+ */
+static void test_fr_pair_afrom_da(void)
+{
+	fr_pair_t   *vp;
+
+	TEST_CASE("Allocation using fr_pair_afrom_da");
+	TEST_CHECK((vp = fr_pair_afrom_da(autofree, attr_test_integer)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	vp->vp_uint32 = 12345;
+
+	TEST_CHECK(vp->vp_uint32 == 12345);
+	TEST_MSG("Expected %s == 12345", attr_test_integer->name);
+
+	talloc_free(vp);
+}
+
+static void test_fr_pair_afrom_child_num(void)
+{
+	fr_pair_t    *vp;
+	unsigned int attr = FR_MODULE_RETURN_CODE;
+
+	TEST_CASE("Allocation using fr_pair_afrom_child_num");
+	TEST_CHECK((vp = fr_pair_afrom_child_num(autofree, fr_dict_root(dict_test), attr)) != NULL);
+
+	TEST_CASE("Validating VP_VERIFY()");
+	VP_VERIFY(vp);
+
+	TEST_CHECK(vp && vp->da && vp->da->attr == FR_MODULE_RETURN_CODE);
```

Comment:
Expressions like this in test cases are really annoying because you have no idea which term evaluated to false.  each term should be wrapped in its own `TEST_CHECK` macro call.  Then the test framework tells you which one isn't true.... Though you don't really need to check vp and vp->da because `VP_VERIFY` already checks that.